src/inferencer/constraintStore.ts

```diff
@@ -24,7 +32,7 @@ function solveConstraint(constraintLhs: Type, constraintRhs: Type): any | undefi
   console.log(toPrint)
   // logging - end
 
-  // check if both key and value are base types and of the same kind (Rule 1)
+  // check if both sides are base types and of the same kind (Rule 1)
   if (
```

Comment:
For all the if else clauses you check if `constraintLhs !== undefined` and `constraintRhs !== undefined`. 
What could be done is to put it all in one clause like this:
```
if (constraintLhs === undefined || constraintRhs === undefined) {
    // console.log the error message
}
```
Before doing all the constraint solving.